gunrock/app/sm/sm_enactor.cuh

```diff
@@ -69,44 +70,174 @@ struct SMIterationLoop : public IterationLoopBase
+            const VertexT &src, VertexT &dest, const SizeT &edge_id,
+            const VertexT &input_item, const SizeT &input_pos,
+            SizeT &output_pos) -> bool
+        {
+            if (src < 0 || src >= nodes_data)
+                return false;
+            if ((!isValid[src]) || (!isValid[dest])) {
+                return false;
+            }
+            // NG has query node id sequence in its even pos; min degree of neighbors in odd pos
+            VertexT query_id = NG[counter[0] * 2];
+            // special init for first iteration
+            if (counter[0] == 0) {
+                if (subgraphs[src] < (query_ro[query_id + 1] - query_ro[query_id]))
+                    return false;
+                // 1 way look ahead
+                if (subgraphs[dest] < NG[counter[0] * 2 + 1]) {
+                    return false;
+                }
+                flags[src] = true;
+                return true;
+            } else {
+                // check if src belongs to partial results
+                if (!flags[src])
+                    return false;
+                if (subgraphs[dest] < (query_ro[query_id + 1] - query_ro[query_id]))
+                    return false;
+                // 1 way look ahead to be done in filter
+		results[dest] = 1;
+                return true;
+            }
+        };
+        auto look_ahead_op = [isValid, flags, nodes_data] __host__ __device__(
+            const VertexT &src, VertexT &dest, const SizeT &edge_id,
+            const VertexT &input_item, const SizeT &input_pos,
+            SizeT &output_pos) -> bool
+        {
+            if (src < 0 || src >= nodes_data)
+                return false;
+            if ((!isValid[src]) || (!isValid[dest])) {
+                return false;
+            }
+            flags[src] = true;
+            return true;
+        };
         // Compute number of triangles for each edge and atomicly add the count to each node, then divided by 2
         // The intersection operation
         auto intersect_op = [subgraphs] __host__ __device__(
             VertexT &comm_node, VertexT &edge) -> bool
         {
             atomicAdd(subgraphs + comm_node,  1);
-            
             return true;
         };
+
         frontier.queue_length = graph.edges;
-        frontier.queue_reset  = true;
-        GUARD_CU(oprtr::Intersect<oprtr::OprtrType_V2V>(
-            graph.csr(), frontier.V_Q(), frontier.Next_V_Q(), 
-            oprtr_parameters, intersect_op));
-
-	// // Sort the subgraph matching values for each node in descending order
- //        GUARD_CU(util::cubSortPairs(
- //            cub_temp_space,
- //            subgraphs, tats1,
- //            nodes,      nodes1,
- //            graph.nodes, 0, sizeof(ValueT) * 8, stream));
+        frontier.queue_reset = true;
+        size_t pointer_head = 0;
+        int num_init = 3; // for current tests, 3 is enough
```

Comment:
How is this determined? It needs to be a command line parameter if given a different dataset we need to update this, won't make sense to recompile every time.

Reply:
Me:
> How big can the query graph get?
Leyuan:
> It's fairly small.
> Usually 3-6 nodes.

Reply:
3-6 nodes is pretty small. What happens if the query graph gets bigger? Does it scale well? Poorly? Or break? How big are the query graphs in the literature / in GraphChallenge?